### DISCUSSION TITLE
IDP linking existing accounts - seems to rely on cookies or cached information and does not actually fully confirm

### DISCUSSION THREAD

**Discussion by Ketec** [keycloak/keycloak#31810] [Q&A]
User has account A (no password), they log in through IDP B.
They are now asked to confirm linking the account by email (no password to login-verify).
They get the email, click the link and see a successful message.

They reload the page, try to login and get the "confirm linking" page again.

It seems that unless you open the email link in the exact same browser instance, it does not register.
Another browser, phone, mail app etc.

They have to click the "continue" button manually - but 99% of the users do not understand that.

Why can't i tbehave like every other normal account confirmation that finishes confirmation automatically once you use the provided link?
Why the extra step of a manual button click (it clearly remembered that link was clicked)?

**Comment by darius-m** [comment]
If I understand it correctly, Keycloak basically requires users to either be authenticated (in the browser that opens the confirmation link), or click a button to confirm that they intend to link the email to the Keycloak account. This approach does make sense, since Keycloak is a SSO and it must make sure that the information it provides is as trustworthy as possible.

Consider the following scenario: you have a client application that uses the email in the SSO response to identify users. One of the users changes their email to `admin@example.com`, which is the email associated with one of the administrator accounts of the client. When that happens, an email is sent to that email address, and someone who has access to the email address either clicks the link by mistake, or is convinced to do so by various means (e.g., phishing attempt). If Keycloak does not attempt to ensure that the user either has access to the account, or in case they are in a different browser, confirm they intentionally clicked the link, the attacker now has the admin email address associated with their account. Because of this, they will get administrator access in the client application on the next login. This is not really a hypothetical, some applications enforce sending emails when linking accounts with SSOs for this specific reason (i.e., they do not trust that the SSO correctly enforces email verifications).

**Comment by Ketec** [comment]
But it is not very functional system for users. Majority do not understand that, They open email in their mail client or phone, it doesn't work.
They instantly also get another email.

It's not even a different confirm page - it just ends up on the same "sent email" page that automatically sends a new email.
With the "continue" link at the bottom of the text block.
